### PR TITLE
[scripts|msbuild] pass toolchain flags

### PR DESCRIPTION
**Pull request by Neumann-A**
Ports using msbuild:

~~chakracore~~
zeroc-ice

~~3fd~~
~~ace~~
~~activemq-cpp~~
~~easyhook~~
~~ecsutil~~
~~gsoap~~
~~hunspell~~ -> moved to make
~~ideviceinstaller~~
~~idevicerestore~~
~~libfabric~~
~~libgpg-error~~
~~libirecovery~~
~~libmicrohttpd~~
~~libosip2~~
~~libusb~~
~~libvpx~~
~~mdnsresponder~~
~~monkeys-audio~~
~~mp3lame~~
~~mpir~~
~~ngspice~~
~~nmap~~
~~ocilib~~
~~openni2~~
~~pbc~~
~~pmdk~~
~~python2~~
~~python3~~
~~rsasynccpp~~
~~scintilla~~
~~sdl1~~
~~tbb~~ -> moved to cmake
~~tracy~~ -> done by somebody else
~~unicorn~~
~~unrar~~
~~usbmuxd~~
~~winpcap~~

### CODE REVIEW

**Comment by Neumann-A** [comment]
ACE looks like supports `"!uwp"`.

**Comment by Neumann-A** [comment]
@BillyONeal Are the releases of the WindowsSDK also gated by vcpkg? The errors I see here in ecsutil are related to https://developercommunity.visualstudio.com/t/code-using-windows-10-sdk-100203480-cant-be-compil/1521695

**Comment by BillyONeal** [comment]
>@BillyONeal Are the releases of the WindowsSDK also gated by vcpkg?

I believe we are currently installing 22621 in our lab: https://github.com/microsoft/vcpkg/blob/a9a23172a7768a05e5d5763e5dee8658eae1bdf9/scripts/azure-pipelines/windows/deploy-visual-studio.ps1#L21

**Comment by Neumann-A** [comment]
hmm don't know what to do with the arm chakracore error. 
the unicorn regressions seems to be a bug in vcpkg-tool.

**Comment by Neumann-A** [comment]
@JackBoosY readosm seems to be a baseline regression.

File: ports/vcpkg-msbuild/vcpkg_msbuild_install.cmake [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+            endif()
+            if(NOT dlls STREQUAL "")
+                file(COPY ${dlls} DESTINATION "${CURRENT_PACKAGES_DIR}/debug/bin")
+            endif()
+        endif()
+    endif()
+
+    vcpkg_copy_pdbs()
+
+    if(arg_CLEAN)
```
Comment (JackBoosY) [inline ports/vcpkg-msbuild/vcpkg_msbuild_install.cmake:151]: Since `vcpkg_install_msbuild` is called 30 times but `SKIP_CLEAN` is only called 7 times, should we set this as a default option?

File: versions/l-/libgpg-error.json [hunk 2]
```diff
@@ -1,7 +1,7 @@
 {
   "versions": [
     {
-      "git-tree": "728a8999d1083dc72eae0612669f0c34075a3f01",
+      "git-tree": "8cc6a9c3aa56443bf6ad9a822f880ae52cdd86a6",
```
Comment (JackBoosY) [inline versions/l-/libgpg-error.json:4]: This is incorrect, we should add a new one.

File: ports/vcpkg-msbuild/vcpkg.json [hunk 3]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "vcpkg-msbuild",
+  "version-date": "2022-08-04",
+  "port-version": 1,
+  "documentation": "https://vcpkg.io/en/docs/README.html",
+  "license": "MIT",
+  "supports": "native",
```
Comment (JackBoosY) [inline ports/vcpkg-msbuild/vcpkg.json:7]:
```suggestion
  "supports": "windows",
```

File: ports/libgpg-error/vcpkg.json [hunk 4]
```diff
@@ -4,5 +4,12 @@
   "port-version": 4,
```
Comment (JackBoosY) [inline ports/libgpg-error/vcpkg.json:4]:
```suggestion
  "port-version": 5,
```

File: ports/ace/vcpkg.json [hunk 5]
```diff
@@ -1,10 +1,19 @@
 {
   "name": "ace",
   "version": "7.0.8",
+  "port-version": 1,
```
Comment (JackBoosY) [inline ports/ace/vcpkg.json:4]: Maybe we can cleanup ci.baseline.txt?

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: ports/python2/portfile.cmake [hunk 6]
```diff
@@ -79,15 +80,19 @@ if (VCPKG_TARGET_IS_WINDOWS)
     
     file(COPY ${SOURCE_PATH}/Lib DESTINATION ${CURRENT_PACKAGES_DIR}/share/python${PYTHON_VERSION_MAJOR})
     
-    file(COPY ${SOURCE_PATH}/PCBuild/${OUT_DIR}/python${PYTHON_VERSION_MAJOR}${PYTHON_VERSION_MINOR}.lib DESTINATION ${CURRENT_PACKAGES_DIR}/lib)
-    file(COPY ${SOURCE_PATH}/PCBuild/${OUT_DIR}/python${PYTHON_VERSION_MAJOR}${PYTHON_VERSION_MINOR}_d.lib DESTINATION ${CURRENT_PACKAGES_DIR}/debug/lib)
+    file(COPY ${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-rel/PCBuild/${OUT_DIR}/python${PYTHON_VERSION_MAJOR}${PYTHON_VERSION_MINOR}.lib DESTINATION ${CURRENT_PACKAGES_DIR}/lib)
```
Comment (Osyotr) [inline ports/python2/portfile.cmake:83]: Quotes here and below

File: ports/python2/portfile.cmake [hunk 7]
```diff
@@ -67,8 +67,9 @@ if (VCPKG_TARGET_IS_WINDOWS)
         message(FATAL_ERROR "Unsupported architecture: ${VCPKG_TARGET_ARCHITECTURE}")
     endif()
     
-    vcpkg_build_msbuild(
-        PROJECT_PATH ${SOURCE_PATH}/PCBuild/pythoncore.vcxproj
+    vcpkg_msbuild_install(
+        SOURCE_PATH "${SOURCE_PATH}"
+        PROJECT_SUBPATH "PCBuild/pythoncore.vcxproj"
```
Comment (Osyotr) [inline ports/python2/portfile.cmake:72]: This should probably build python.sln instead to match Linux version, but I think this change belongs to a separate PR.

File: ports/python3/use_diff_makro.patch [hunk 8]
```diff
@@ -0,0 +1,13 @@
+diff --git a/PC/launcher.c b/PC/launcher.c
+index 734e75333..7124927cf 100644
+--- a/PC/launcher.c
++++ b/PC/launcher.c
+@@ -2015,7 +2015,7 @@ installed, use -0 for available pythons", &p[1]);
+     return rc;
+ }
+ 
+-#if defined(_WINDOWS)
++#if !defined(_CONSOLE) // _WINDOWS is defined by the vcpkg toolchain
```
Comment (Osyotr) [inline ports/python3/use_diff_makro.patch:10]: `#if defined(WIN32)`?

File: ports/vcpkg-msbuild/vcpkg.json (hunk 3 above)
Comment (Neumann-A) [reply ports/vcpkg-msbuild/vcpkg.json:7]: that is windows&native

File: ports/python3/use_diff_makro.patch (hunk 8 above)
Comment (Neumann-A) [reply ports/python3/use_diff_makro.patch:10]: No this is a switch between `WinMain` and `wmain`.

File: ports/ace/vcpkg.json (hunk 5 above)
Comment (Neumann-A) [reply ports/ace/vcpkg.json:4]: comment on the supports line ? (this is port-version)

File: versions/l-/libgpg-error.json (hunk 2 above)
Comment (Neumann-A) [reply versions/l-/libgpg-error.json:4]: seems like a merge messed it up

File: ports/python2/portfile.cmake (hunk 7 above)
Comment (Osyotr) [reply ports/python2/portfile.cmake:72]: I mean, on windows this port produces only library, but on Linux it also produces the tool.

Comment (Neumann-A) [reply ports/python2/portfile.cmake:72]: ah ok. Yeah that is a different PR

File: ports/vcpkg-msbuild/vcpkg_msbuild_install.cmake (hunk 1 above)
Comment (Neumann-A) [reply ports/vcpkg-msbuild/vcpkg_msbuild_install.cmake:151]: vcpkg_msbuild_install copys the sources into the separate build folders so there is no reason to every call the clean target.

**Review by JackBoosY — DISMISSED** [review]
Needs other guys review vcpkg-msbuild part.

**Comment by Neumann-A** [comment]
since those conflicts are only v db conflicts I'll fix them after review.

**Comment by Neumann-A** [comment]
@BillyONeal Should I split this PR in only the `vcpkg-msbuild` port and then silently update the other ports into using it? Would make that merging this a lot faster? (I mean this PR already proved that it will work so splitting it out should be a thing.)

**Comment by JonLiu1993** [comment]
@Neumann-A, according to your PR, I split `vcpkg-msbuild` and `vcpkg-pkgconfig-get-modules` and the parts modified by `zeroc-ice` to [test ](https://github.com/microsoft/vcpkg/pull/28740)that the CI baseline error about `zeroc-ice` can be successfully repaired. I would like to ask you to put the most in this PR The basic part involves `vcpkg-msbuild` and the part of `vcpkg-pkgconfig-get-modules re-split` Submit a PR and then we can fix the CI baseline error about `zeroc-ice` as soon as possible.

File: ports/vcpkg-pkgconfig-get-modules/x_vcpkg_pkgconfig_get_modules.cmake [hunk 9]
```diff
@@ -11,6 +11,10 @@ function(x_vcpkg_pkgconfig_get_modules)
     if(DEFINED arg_UNPARSED_ARGUMENTS)
         message(FATAL_ERROR "extra arguments passed to x_vcpkg_pkgconfig_get_modules: ${arg_UNPARSED_ARGUMENTS}")
     endif()
+    set(msvc "")
+    if(arg_MSVC_SYNTAX)
+        set(msvc "--msvc-syntax")
+    endif()
```
Comment (dg0yt) [inline ports/vcpkg-pkgconfig-get-modules/x_vcpkg_pkgconfig_get_modules.cmake:17]: This isn't satisfying yet. I already pass `--msvc-syntax`  with the modules list in some ports. Each of these ports must decide whether it needs to use the option. IMO it should be the default for MSVC triplets, possibly overriden with `MSVC_SYNTAX <bool>`, or with a triplet variable for use in custom triplets.

Comment (Neumann-A) [reply ports/vcpkg-pkgconfig-get-modules/x_vcpkg_pkgconfig_get_modules.cmake:17]: Currently 9 files with x_vcpkg_pkgconfig_get_modules. 4 with `--msvc-syntax`. Looks more like a fifty-fifty for me.

Comment (dg0yt) [reply ports/vcpkg-pkgconfig-get-modules/x_vcpkg_pkgconfig_get_modules.cmake:17]: Well, I have been cautious to not use it too much, given its `x` prefix.
The use cases are:
- `nmake` builds: MSVC syntax
- autotools builds: no MSVC syntax (at least for my coinutils PR).
- `qt5`: I don't remember right now.

**Comment by Neumann-A** [comment]
Converting to draft. This PR is more like a port status PR for the transition as soon as the other PR with the scripts is merged. It also shows how to convert.
